syft/source/directory_resolver.go

```diff
@@ -1,41 +1,211 @@
+	pathFilterFns []pathFilterFn
+	errPaths      map[string]error
+}
+
+func newDirectoryResolver(root string, pathFilters ...pathFilterFn) (*directoryResolver, error) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("could not create directory resolver: %w", err)
+	}
+
+	if pathFilters == nil {
+		pathFilters = []pathFilterFn{isSystemRuntimePath}
+	}
+
+	r := directoryResolver{
+		path:          root,
+		cwd:           cwd,
+		fileTree:      filetree.NewFileTree(),
+		infos:         make(map[file.ID]os.FileInfo),
+		pathFilterFns: pathFilters,
+		errPaths:      make(map[string]error),
+	}
+
+	// why account for multiple roots? To cover cases when there is a symlink that references above the root path,
+	// in which case we need to additionally index where the link resolves to. it's for this reason why the filetree
+	// must be relative to the root of the filesystem (and not just relative to the given path).
+	roots := []string{root}
+loop:
+	for {
+		var p string
+		switch len(roots) {
+		case 0:
+			break loop
+		case 1:
+			p, roots = roots[0], nil
+		default:
+			p, roots = roots[0], roots[1:]
+		}
+
+		additionalRoots, err := r.indexTree(p)
+		if err != nil {
+			return nil, fmt.Errorf("unable to index filesystem: %w", err)
+		}
+		roots = append(roots, additionalRoots...)
+	}
+
+	return &r, nil
+}
+
+func (r *directoryResolver) indexTree(root string) ([]string, error) {
+	log.Infof("indexing filesystem path=%q", root)
+	var err error
+	root, err = filepath.Abs(root)
+	if err != nil {
+		return nil, err
+	}
+	var roots []string
+	stager, prog := indexingProgress(root)
+	defer prog.SetCompleted()
+
+	return roots, filepath.Walk(root,
+		func(p string, info os.FileInfo, err error) error {
+			stager.Current = p
+
+			// ignore any path which a filter function returns true
+			for _, fn := range r.pathFilterFns {
+				if fn(p) {
+					return nil
+				}
+			}
+
+			// permission denied, IO error, etc... we keep track of the paths we can't see, but continue with indexing
+			if err != nil {
+				log.Warnf("unable to index path=%q: %+v", p, err)
+				r.errPaths[p] = err
+				return nil
+			}
+
+			// link cycles could cause a revisit --we should not allow this
+			if r.fileTree.HasPath(file.Path(p)) {
+				return nil
+			}
+
+			newRoot, err := r.addPathToIndex(p, info)
+			if err != nil {
+				return err
+			}
+
+			if newRoot != "" {
+				roots = append(roots, newRoot)
+			}
+
+			return nil
+		})
+}
+
+func (r directoryResolver) addPathToIndex(p string, info os.FileInfo) (string, error) {
+	var ref *file.Reference
+	var err error
+	var newRoot string
+
+	switch newFileTypeFromMode(info.Mode()) {
+	case SymbolicLink:
+		linkTarget, err := os.Readlink(p)
+		if err != nil {
+			if errors.Is(err, os.ErrPermission) {
+				// don't allow for permission errors to stop indexing, keep track of the paths and continue.
+				log.Warnf("unable to index symlink=%q: %+v", p, err)
+				r.errPaths[p] = err
+				return "", nil
```

Comment:
would prefer some consistency here. perhaps we can return an error that clearly states it's a permission failure, and filter it in the calling method `indexTree`. That way, it's not implicit in the return value being empty. ex.
```
return "", PermissionError("unable to index symlink")
...

// caller
newRoot, err := r.addPathToIndex(path, info)
if err != nil && err != PermissionError {
  return err
}
roots = append(roots, newRoot)
```

Reply:
handled in a separate function as well as migrated handling to the caller